### PR TITLE
Complete rewrite from scratch is here \o/

### LINKED ISSUE #49: Error upgrading to v3.2.0
I didn't make any change in my code. Just updated package from 3.1.3 to 3.2.0 and I'm getting this error at Bowser Console.

```
app.js:10124 Warning: React.createElement: type is invalid -- expected a string (for built-in components) or a class/function (for composite components) but got: undefined. You likely forgot to export your component from the file its defined in. Check the render method of If.
    in If (created by PageFooter)
    in div (created by PageFooter)
    in div (created by PageFooter)
    in footer (created by PageFooter)
    in PageFooter (created by Connect(PageFooter))
    in Connect(PageFooter) (created by App)
    in div (created by App)
    in div (created by App)
    in DocumentTitle (created by SideEffect(DocumentTitle))
    in SideEffect(DocumentTitle) (created by App)
    in App (created by Connect(App))
    in Connect(App)
    in Provider

app.js:10294 Uncaught Invariant Violation: Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: undefined. You likely forgot to export your component from the file its defined in. Check the render method of If.
    at invariant (http://localhost:5000/app.js:10294:16)
    at ReactCompositeComponentWrapper.instantiateReactComponent [as _instantiateReactComponent] (http://localhost:5000/app.js:25172:56)
    at ReactCompositeComponentWrapper.performInitialMount (http://localhost:5000/app.js:25596:23)
    at ReactCompositeComponentWrapper.mountComponent (http://localhost:5000/app.js:25487:22)
    at Object.mountComponent (http://localhost:5000/app.js:17982:36)
    at ReactDOMComponent.mountChildren (http://localhost:5000/app.js:24634:45)
    at ReactDOMComponent._createInitialChildren (http://localhost:5000/app.js:21718:33)
    at ReactDOMComponent.mountComponent (http://localhost:5000/app.js:21537:13)
    at Object.mountComponent (http://localhost:5000/app.js:17982:36)
    at ReactDOMComponent.mountChildren (http://localhost:5000/app.js:24634:45)

app.js:14605 Uncaught TypeError: Cannot read property __reactInternalInstance$t7ajv2eucf of null
    at Object.getClosestInstanceFromNode (app.js:14605)
    at findParent (app.js:28238)
    at handleTopLevelImpl (app.js:28267)
    at ReactDefaultBatchingStrategyTransaction.perform (app.js:18964)
    at Object.batchedUpdates (app.js:28192)
    at Object.batchedUpdates (app.js:17634)
    at dispatchEvent (app.js:28347)
```

Could this be a conflict with other outdated package? I'm using react 15.6.2.

### LINKED ISSUE #55: Allow non-booleans in the `condition` prop as a convenience
I see a lot of these warnings:

```
Warning: Failed prop type: Invalid prop `children` supplied to `If`.
```

Because I will have conditions that look like:

```jsx
<When condition={state.data && state.data.myProfile}>
  My profile loaded!
</When>
```

With `state.data.myProfile` being an object that might look like:

```json
{
  "id": 123,
  "name": "Joe",
  "email": "charley@email.com"
}
```

Which means that I have to do this to silence warnings:

```jsx
<When condition={!!(state.data && state.data.myProfile)}>
  My profile loaded!
</When>
```

Or:

```jsx
<When condition={state.data && state.data.myProfile && true}>
  My profile loaded!
</When>
```

Both of these solutions seems solvable if `react-if` accepted non-boolean conditions.

`null` and `undefined`  or anything non-truthy would be evaluated to a `false`.

### PR DESCRIPTION
**Pull request by Favna**
This is a complete rewrite of the lib from the ground up. Using the original code as a basis while completely reimplementing everything.

fixes #55 closes #57

closes #60 closes #62 closes #49

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+  extends: ['@favware/eslint-config-react-app', 'plugin:prettier/recommended'],
+  overrides: [
+    {
+      files: ['./tsdx.config.js'],
+      rules: {
+        '@typescript-eslint/no-var-requires': 0
+      }
+    }
+  ]
+};
```
Comment (cfanoulis) [inline .eslintrc.js:11]: why JS tho... JSON was alright

File: README.md [hunk 2]
```diff
@@ -1,11 +1,8 @@
-
 # React If
 
 [![npm badge](http://img.shields.io/npm/v/react-if.svg)](https://www.npmjs.com/package/react-if)
-[![Build Status](https://travis-ci.org/romac/react-if.svg?branch=master&style=flat)](https://travis-ci.org/romac/react-if)
```
Comment (cfanoulis) [inline README.md:5]: please reimplement for github actions kthx

File: demo/README.md [hunk 3]
```diff
@@ -1,8 +0,0 @@
-
-# Demo
-
-- `npm install --dev`
-- `npm run build`
-- open `index.html` in the browser of your choice
-
-
```
Comment (cfanoulis) [inline demo/README.md:8]: undelete this pls

File: demo/tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    "jsx": "react",
+    "moduleResolution": "node",
+    "noImplicitAny": false,
+    "noUnusedLocals": false,
+    "noUnusedParameters": false,
+    "removeComments": true,
+    "strictNullChecks": true,
+    "preserveConstEnums": true,
+    "sourceMap": true,
+    "lib": ["es2015", "es2016", "dom"],
```
Comment (cfanoulis) [inline demo/tsconfig.json:15]: newer lib maybe? es2015 died with node8 no?

**Review by cfanoulis — COMMENTED** [review]
Hello! Just a couple docs/tool nits, the code looks well

File: .eslintrc.js (hunk 1 above)
Comment (Favna) [reply .eslintrc.js]: Covered in df46295605c9b177909666a56225216ca6c7e1b0

File: README.md (hunk 2 above)
Comment (Favna) [reply README.md:5]: Covered in 2109d182ed23c84c46d8d3a0c39ce24d6b81857d

File: demo/README.md (hunk 3 above)
Comment (Favna) [reply demo/README.md:8]: Covered in c28bd847317f9b5c7eb0240b57da6392671144a7

File: demo/tsconfig.json (hunk 4 above)
Comment (Favna) [reply demo/tsconfig.json]: Covered in d613ee14515bf6c12bcc06051cbcbcc6260e261f

**Comment by Favna** [comment]
Note: Size Limit Analysis CI is failing because there is no size-limit on master branch yet. Ignore it for now.

**Review by cfanoulis — APPROVED** [review]

**Pull request merged**